### ISSUE TITLE
🐛 Self-closing tags do not exist in HTML

### ISSUE THREAD

**Issue by shadow-identity** [biomejs/biome#4490]
### Environment information

```block
I don't have it locally, tried in playground instead.
https://biomejs.dev/playground/?enabledLinting=false&files.main.html=PABpAG4AcAB1AHQAPgAKADwAaQBtAGcAPgAKADwAYgByAD4ACgA%3D
```

### What happened?

1. Create valid .html file with [void elements](https://developer.mozilla.org/en-US/docs/Glossary/Void_element).
2. Disable JSX syntax support in Biome
3. Format the file with Biome
4. Observe that Biome corrupted valid void HTML elements by making them self-closing.

### Expected result

Since there are no self-closing tags in HMTL, Biome should keep valid void elements formatting.
Leading slash should be added only where self-closing tags are valid: within XML-like content such as svg, xml or JSX.

For these who want to keep having broken Prettier formatting of void HTML tags, there could be a configuration option.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
FYI Biome [**doesn't support HTML** yet](https://biomejs.dev/internals/language-support/). The playground allows HTML because we are adding its support via an experimental flag while we work on it.

We added the `/` because we follow Prettier formatting. What should we do instead? 

This is about aligning with Prettier. What's your suggestion?

**Comment by dyc3** [comment]
Biome's objective is to align with prettier's formatting. You are right in that technically self closing tags don't exist in HTML. However, they do exist in HTML-ish templating languages, which we do plan to support. Regardless of correctness, we still need to handle self closing syntax because its a frequently used pattern.

The behavior displayed in your playground is intentional and not a bug. Biome correctly determines that these are void elements and uses self-closing syntax when formatting.

**Comment by ematipico** [comment]
> > because we follow Prettier
> 
> When formatting HTML, you don't need to follow Prettier, you need to follow HTML.

@firefoxic  Thank you; however, what do you suggest? I get the sense you would remove the `/`, but would you provide an option to align with Prettier?

**Comment by shadow-identity** [comment]
> This is about aligning with Prettier. What's your suggestion?

That's really cool, because it allows to easily switch from Prettier to Biome in the future (and I'd do it!).

Nevertheless, please consider that there are Prettier plugins aimed to fix Prettier's void tag corruption like [prettier-plugin-void-html](https://www.npmjs.com/package/@awmottaz/prettier-plugin-void-html). And if we use the plugin now, and we switch to Biome, our code will break.

Also, the wide presence of Prettier made a lot of new-comming front end developers to think that this syntax is valid.

So my suggestion would be to make this behavior configurable. So people who want to seamlessly migrate from Prettier can simply turn it on. The rest will have the valid HTML in their projects by default. And these who just join the front-end community using Biome will not be confused.

**Comment by pepelsbey** [comment]
> Biome's objective is to align with prettier's formatting

While aligning with Prettier, you can make it even better: offer an option similar to what [prettier-plugin-void-html](https://www.npmjs.com/package/@awmottaz/prettier-plugin-void-html) plugin does, but natively. I’d prefer Biome to be HTML-compatible by default and optionally Prettier-compatible, but that’s just me :)

**Comment by egrieco** [comment]
> I’d prefer Biome to be HTML-compatible by default and optionally Prettier-compatible, but that’s just me :)

It's not just you, many of us are baffled at Prettier's obstinate refusal to emit correct HTML code, and instead force XHTML syntax.

Being opinionated is okay when there are options, but when a tool is contradicting standards it's obnoxious and inarguably wrong.

**Comment by bavalpey** [comment]
Hi Biome, please consider diverging from prettier here, as not adhering to correct HTML code is unacceptable.

**Comment by minht11** [comment]
Svelte had similar discussion and debate about self [closing tags](https://github.com/sveltejs/svelte/issues/11052). Ultimately they decided to follow HTML spec.

In my personal view while self closing tags _look nice_ they are wrong when used in HTML and produce different behavior than users would actually expect. I feel like correctness should win here.

Prettier [discussion](https://github.com/prettier/prettier/issues/5246)

**Comment by ematipico** [comment]
The thing is, self-closing tags don't exist in _HTML_, but they **are required** for void elements in XML, XHTML and SVG.

Which means, if a `.html` file adds the XHTML doc type metadata*, the self closing tag on void elements becomes mandatory.

https://developer.mozilla.org/en-US/docs/Glossary/Void_element#self-closing_tags

*And that's when the "issue" arises, knowing when a `.html` file is an actual XHTML file spec, which can only be inferred by the metadata. Maybe prettier went for the convenient way, to add a self-closing tag every time. It doesn't harm HTML files, and XHTML files don't have parsing errors

**Comment by pepelsbey** [comment]
> The thing is, self-closing tags don't exist in HTML, but they are required for void elements in XML, XHTML and SVG.

Fortunately, XHTML doesn't exist on the web. You need to work really hard to force browsers to parse your markup as XHTML. And then you'd end up breaking the whole page in a parsing error when a stray ampersand, missing closing tag, or quote is detected. Nobody wants that. So, this concern is purely theoretical.

As for the XML-based languages you can embed into HTML, like SVG or MathML, they have their own parsing rules that browsers apply when they detect `<svg>` or `<math>` elements. Disregarding the difference in parsing languages is not a very smart thing to do. Yes, they look the same, but they work differently. You wouldn't format CSS inside of `<style>` or JS inside of `<script>` as HTML just because they're in the .html files, wouldn't you?

I think language formatters should accept the reality of how languages and browsers work, not invent it.

**Comment by minht11** [comment]
There are two different cases we are conflating a bit:

- Self closing tags on void elements, those do nothing and does not change result.
- Self closing tags on regular elements, this is where problem happens.

Take input like this:
```html
<div />
Content <button>
 My button
</button>
```
If you use JSX, you would expect output to be:
```html
<div></div>
Content
<button>
 My button
</button>
```
but [browser actually](https://jsfiddle.net/my80g6wo/1/) parses this as:
```html
<div>
  Content
  <button>
   My button
  </button>
</div>
```
Prettier happily leaves this as is, even though in HTML this produces unexpected results. Even Svelte creator was surprised to learn about this behavior last year, so regular users won't do any better.

**Comment by bavalpey** [comment]
> Self closing tags on void elements, those do nothing and does not change result.

This is incorrect.
This violates the spec, meaning it violated correctness, meaning that browsers are able to do whatever they want when they see a self closing tag. You can't violate the rules of a spec document. The result can change at literally any time.
This also works against smaller browsers that don't all parse code the same way. Imagine you're a tinkerer writing your own browser renderer. You'd follow the spec.


*All code formatted by prettier that has a void element will be reported as invalid according to [W3C's HTML validator](https://validator.w3.org/). This is unacceptable.


Lots of browsers will display my page in many cases if I forget to close `<body>`, not changing the visual result. Or if I add an extra closing tag that wasn't opened.

I can't tell from your response and the rest of the thread responses whether you plan on correcting prettier's wrong, but I sure hope you do. Don't follow Prettier's blatant arrogance. Respect your developers and the HTML specification.

**Comment by QuentinRoy** [comment]
It does not, actually, violate the specs. The specs say to ignore it.

**Comment by pepelsbey** [comment]
> Browsers are able to do whatever they want when they see a self closing tag. You can't violate the rules of a spec document. The result can change at literally any time.

Not exactly: the HTML spec has detailed algorithms for dealing with incorrect markup, and all browser parsers are based on that. So, whatever you do, all browsers will give you the same result. But most of the developers have no idea how complex those algorithms are. I wrote an [article about jumping HTML tags](https://pepelsbey.dev/articles/jumping-html-tags/) you might find interesting.

**Comment by bavalpey** [comment]
> However it does not, actually, violate the specs. The specs specifically say to ignore the extra slash.

[This is incorrect](https://github.com/validator/validator/wiki/Markup-%C2%BB-Void-elements#trailing-slashes-directly-preceded-by-unquoted-attribute-values). The HTML spec advises against using them. In other words, it is against the spec. At best, they are ignored. At worst, they change the expected meaning.

**Comment by dyc3** [comment]
I'm open to adding an HTML specific option to diverge from Prettier here, so let's do it. I'm also open to suggestions for naming the option and the values for this new option.

I would like to note that this new option would need to **only apply to HTML**, because self-closing tags are a feature of some HTML-ish templating languages, like Vue, Svelte, and others.

**Comment by bavalpey** [comment]
@dyc3 

Thank you for the consideration.

Also, sorry, the option shouldn't be to diverge from prettier.
The option should be to diverge from the spec.

In other words, the default should emphatically *not* be what prettier does.

Call the option `no-html-void-tags` or `prettier-style-void-tags` or something.

**Comment by minht11** [comment]
> because self-closing tags are a feature of some HTML-ish templating languages, like Vue, Svelte, and others

Svelte does not support self closing tags for regular HTML elements as of [Svelte 5.0](https://github.com/sveltejs/svelte/issues/11052).
Components do support self closing tags tho.

**Comment by ematipico** [comment]
> This violates the spec, meaning it violated correctness, meaning that browsers are able to do whatever they want when they see a self closing tag. You can't violate the rules of a spec document. The result can change at literally any time.

Do you have any resources that talk about this?

**Comment by bavalpey** [comment]
> > This violates the spec, meaning it violated correctness, meaning that browsers are able to do whatever they want when they see a self closing tag. You can't violate the rules of a spec document. The result can change at literally any time.
> 
> Do you have any resources that talk about this?

Resources that talk about the spec? I already linked the one. I can link the actual spec, too: https://html.spec.whatwg.org/#start-tags

It permits them, but recommends against using them.

As for the rest of that? This is just from common knowledge of being a compiler developer. Your only contract is the spec. That's it.

**Comment by egrieco** [comment]
> [@dyc3](https://github.com/dyc3)
> 
> Also, sorry, the option shouldn't be to diverge from prettier. The option should be to diverge from the spec.
> 
> In other words, the default should emphatically _not_ be what prettier does.
> 
> Call the option `no-html-void-tags` or `prettier-style-void-tags` or something.
> 
> This is because otherwise you are encouraging and misinforming new and novice web developers.

Doing it the "prettier" i.e. incorrect way contradicts Biome's stated goal:

> [Fix problems, learn best practice](https://biomejs.dev/#:~:text=Fix%20problems%2C%20learn%20best%20practice)

**Comment by 43081j** [comment]
its unfortunate you've hit this landmine by trying to align with prettier. the discussion over there on this very same topic has been going on for 7+ years now.

all of it drills down to the fact that HTML certainly should not have self-closing tags printed, but non-HTML (though HTML-_like_) template languages may.

JSX and vue in particular use self-closing tags. so your printer should treat them differently to HTML.

this shouldn't really be a configuration option, it should be a conditional behaviour based on the language you're formatting. the mistake here is treating them all as if they're HTML when they're not.

**Comment by ematipico** [comment]
I had a small chat on Discord with the other maintainers, and they generally agree to the following plan: 

> - parser: it does parse void elements and self-closing elements (e.g. <input /> and <input > are both allowed)
> - parser: if it parses `.html` files, it should raise a diagnostics if it sees a self-closing element. Maybe a warning? I can't recall if our parser infra can raise *warnings*, but we could make our parser stricter via options.
> - formatter, **by default** removes the self-closing tag `<input />` -> `<input >` inside .html files
> - formatter will expose an option for prettier compatibility: it adds the / e.g. `<input >` -> `<input />` in .html files
>
> I would completely ignore XHTML for now. Who uses it in a modern setup, nowadays?

___

> its unfortunate you've hit this landmine by trying to align with prettier. the discussion over there on this very same topic has been going on for 7+ years now.

I don't deny it, and I also think that having Prettier compatibility is what has made Biome what it is. 

>  it should be a conditional behaviour based on the language you're formatting

Yes, that's the plan. The option is mostly to make the transition from Prettier easier for users. We can bikeshed the name and semantically explain it as a divergence from the spec. While all the arguments of this issue are valid, if Biome makes the transition as painless as possible (effort, resources, code review times, etc.), it's a big win for Biome as a project.
